utilities/env_librados.cc

```diff
@@ -0,0 +1,1043 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "rocksdb/utilities/env_librados.h"
+#include "util/random.h"
+
+#define DEBUG
+
+#ifdef DEBUG
+#include <cstdio>
+#define LOG_DEBUG(fmt, ...)  do{\
+    printf("[%s:%i:%s]", __FILE__, __LINE__, __FUNCTION__);\
+    printf(fmt, __VA_ARGS__);\
+    }while(0)
+#else
+#define LOG_DEBUG(fmt, ...)
+#endif
+
+
+/* GLOBAL CONSTANT */
+const std::string DIR_TAG_VALUE = "<dir>";
+const int MAX_ITEMS_IN_FS = 1<<20;
+const std::string TEST_DIR_PREFIX = "TEST_TEMP_";
+const std::string ROOT_DIR_TAG = "/";
+
+rocksdb::Status err_to_status(int r)
+{
+  switch (r) {
+  case 0:
+    return rocksdb::Status::OK();
+  case -ENOENT:
+  case -ENODATA:
+  case -ENOTDIR:
+    return rocksdb::Status::NotFound(rocksdb::Status::kNone);
+  case -EINVAL:
+    return rocksdb::Status::InvalidArgument(rocksdb::Status::kNone);
+  case -EIO:
+    return rocksdb::Status::IOError(rocksdb::Status::kNone);
+  default:
+    // FIXME :(
+    assert(0 == "unrecognized error code");
+    return rocksdb::Status::NotSupported(rocksdb::Status::kNone);
+  }
+}
+
+void split(const std::string &fn, std::string *dir, std::string *file) {
+  size_t slash = fn.rfind('/');
+  *file = fn.substr(slash + 1);
+  while (slash && fn[slash-1] == '/')
+    --slash;
+  *dir = fn.substr(0, slash);
```

Comment:
I think if you change this to make both "/a/b/c" and "b/c" -> dir 'b' and file 'c' it will simplify things.  In practice we can set the db paths to vary only by the final directory component.  rocksdb only really needs a 2-level structure (dir/file) to distinguish between the wal_dir files, regular db files, and any "slow" directories set up via the db_paths option.  We don't need to worry about complicated directory structures...